<a href="https://colab.research.google.com/github/edmenciab733/ecc_qiskit_fallfest/blob/main/operaciones_cuanticas_numpy.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
%%capture
!pip install qiskit
!pip install qiskit-aer

In [2]:
import numpy as np
import qiskit

# Bases Computacional , Base Hadamard, y Base  Y

---

**Base computacional**

- $ |0\rangle $
- $ |1\rangle $

**Operador de Hadamard y transformación a la Base Hadamard**

El operador de Hadamard es una operación unitaria que actúa sobre un único qubit y tiene una variedad de aplicaciones importantes, desde la creación de superposiciones hasta la implementación de algoritmos cuánticos.


- $ H = \frac{1}{\sqrt{2}}\begin{pmatrix} 1 & 1 \\ 1 & -1 \end{pmatrix} $  


- $ |+\rangle $ = $H  |0\rangle $ = $ \frac{|0\rangle + |1\rangle}{\sqrt{2}} $


- $ |-\rangle $= $H  |1\rangle $ = $ \frac{|0\rangle - |1\rangle}{\sqrt{2}} $

**Base Y **

- $ |i\rangle $
- $ |-i\rangle $

## Operadores con numpy

In [3]:
ket_0 = np.array([[1], [0]])
ket_1 = np.array([[0], [1]])

In [4]:
H = 1/np.sqrt(2) * np.array([[1, 1], [1, -1]])

In [5]:
ket_p  = H.dot(ket_0)

In [6]:
ket_m = H.dot(ket_1)

In [7]:
ket_ip =  1/np.sqrt(2) * np.array([[1], [1j]])

In [8]:
ket_im = 1/np.sqrt(2) * np.array([[1], [-1j]])

## Producto Interno
El producto intero, también conocido como producto escalar, es una operación matemática que toma dos vectores en este espacio y devuelve un número complejo. El producto interno se denota como:  $ <v|u >$. Un requísito es aplicar el produnto interno es que el número de columnas del primer vector debe coincidir con el número de filas.

Ejemplo con: $ <0|0> $

$\begin{pmatrix} 1 \\ 0  \end{pmatrix} \times \begin{pmatrix} 1 & 0  \end{pmatrix} $  



In [9]:
ket_0_tc = np.conj(ket_0.T)
ket_0.shape, ket_0_tc.shape

((2, 1), (1, 2))

In [10]:
np.vdot(ket_m, ket_0_tc)

0.7071067811865475

In [11]:
np.vdot(ket_0, ket_0_tc)

1

In [12]:
ket_0_tc = np.conj(ket_0.T)
np.vdot(ket_1, ket_0_tc)

0

El detalle: El producto interno puede darnos un una idea de cuan cerca o lejos están los estados. Probar con: $ <0|0> = 1 $ ó $ <0|1> = 0 $  ó  $ <0|+> = 0.707 $

# Relaciones entre qubits

Las relaciones entre qubits se realizan por medio kronecker

$|00\rangle = |0\rangle \otimes |0\rangle $

$  |01\rangle = |0\rangle \otimes |1\rangle $

$|10\rangle = |1\rangle \otimes |0\rangle $

$|11\rangle = |1\rangle \otimes|1\rangle $




In [13]:
ket_00 = np.kron(ket_0, ket_0)
ket_01 = np.kron(ket_0, ket_1)
ket_10 = np.kron(ket_1, ket_0)
ket_11 = np.kron(ket_1, ket_1)

# Operadores(Puertas) de Pauli, combinaciones y operaciones sobre 1 qubit

Los operadores de Pauli constituyen una serie de matrices que tiene efecto sobre qubits individuales.

La puerta $ X $ es análoga al bit clasico actua sobre el eje $ x $ y su efecto es rotar el eje mencionado un valor $ \pi $

In [14]:
X = np.array([[0,1], [1, 0]])

In [15]:
Z = np.array([[1,0], [0, -1]])

In [16]:
Y = np.array([[0, -1j], [1j, 0]])

In [17]:
I = np.array([[1,0], [0,1]])

In [18]:
np.dot(X, ket_1)

array([[1],
       [0]])

Para combinar dos o n operadores cuanticos que terminaran a afectando a 2 o $ n $ qubits tambien se realiza el producto de kronecker entre los operados por ejemplo para dos puertas $ XX $ sería:

$ XX = X \otimes X $

El <a href="https://repositoriodigital.uns.edu.ar/handle/123456789/5841#:~:text=El%20producto%20de%20Kronecker%20se,el%20c%C3%A1lculo%20de%20valores%20singulares." > producto de kronocker </a>garantiza que los qubits se mantengan independiente. No es una operación sobre dos qubits

In [19]:
XX = np.kron(X, X)
XX

array([[0, 0, 0, 1],
       [0, 0, 1, 0],
       [0, 1, 0, 0],
       [1, 0, 0, 0]])

Ejemplo si queres afectar a aplicar al par de de qubits  $|00\rangle$ con la puerta $ X$ aplicado a ambos qubits. La operación es cuanto sigue:

$ XX |00\rangle $


In [20]:
np.dot(XX, ket_00)

array([[0],
       [0],
       [0],
       [1]])

En la notación de bra-ket, comúnmente utilizada en mecánica cuántica, el producto tensorial de operadores lineales se puede definir de la siguiente manera. Consideremos dos espacios vectoriales, $V$ y $U$ con vectores $ |v\rangle $ en $V$ y $ |u\rangle $ en $U$. El producto tensorial de los operadores  $ A y  B $ denotado por $ A \otimes B $ actúa sobre el espacio producto tensorial de $ U \otimes V $ por tanto se puede decir que tambien actua sobre $ u$  y $ v $






$ (A \otimes B ) ( |v\rangle \otimes |u\rangle ) =  (A \otimes |v\rangle ) (B \otimes |u\rangle)$

Ejemplo

$ ZX|10\rangle = Z|1\rangle X|0\rangle $

In [21]:
ZX = np.kron(Z, X)
ket1ket0 = np.kron(ket_1, ket_0)
ZXket1ket0 = ZX.dot(ket1ket0)
ZXket1ket0

array([[ 0],
       [ 0],
       [ 0],
       [-1]])

In [22]:
Zket1 = Z.dot(ket_1)
Xket0 = X.dot(ket_0)
Zket1Xket0 = np.kron(Zket1,Xket0 )
Zket1Xket0

array([[ 0],
       [ 0],
       [ 0],
       [-1]])

# Puertas RX, RY, RZ o Puertas de desplazamiento de fase.

Las puertas de desplazamiento de fase, o puertas de cambio de fase, en la computación cuántica se utilizan para modificar la fase de un estado cuántico sin cambiar sus probabilidades de medición. Estas puertas son esenciales para controlar y manipular las fases relativas de los estados cuánticos, lo que es crucial para muchos algoritmos cuánticos.


Las puertas desplazamientos no cambia la probabilidad de medir un qubit de manera individual, pero utiliza la propiedad de la interferencia cuantica para incidir en la medida generar del sistema, en el cual se combinan amplitudes de manera constructiva o destructiva. Normalmente inciden en la medida de aquellos estados entrelazados.

$RX(\theta) =  \begin{pmatrix} \cos(\frac{\phi}{2}) & -i * \sin(\frac{\phi}{2}) \\  -i * \sin(\frac{\phi}{2}) & \cos(\frac{\phi}{2}) \end{pmatrix}$

----------
$RZ(\theta) =  \begin{pmatrix} e^{-i\theta/2} & 0 \\
0 & e^{i\theta/2} \end{pmatrix}$

----------

$RY(\theta) =  \begin{pmatrix} \cos(\frac{\phi}{2}) & -\sin(\frac{\phi}{2})  \\
\sin(\frac{\phi}{2})  & \cos(\frac{\phi}{2}) \end{pmatrix}$


In [23]:
def get_matrix_rotacion(puerta, phi=np.pi):
  if puerta == "x":
    return np.array([[np.cos(phi/2), -1j * np.sin(phi/2)], [-1j * np.sin(phi/2), np.cos(phi/2) ]])
  if puerta == "z":
    return np.array([[np.exp(-1j*phi/2), 0], [0 , np.exp(1j*phi/2)  ]])
  if puerta == "y":
    return np.array([[np.cos(phi/2), -1j* np.sin(phi/2)], [np.sin(phi/2) ,np.cos(phi/2)  ]])

In [24]:
rx = get_matrix_rotacion("x")
rz = get_matrix_rotacion("z")
ry = get_matrix_rotacion("y")
aux = rz.dot(rx.dot(ket_0))

La puerta $ RX(\pi) $, aplica una rotación de 180 grados y un cambio en una fase global del sistema. Eso a diferencia de la puerta $X$ que hace el mismo efecto, pero no cambio la fase.

In [26]:
np.round(rx.dot(ket_0), 5)

array([[0.+0.j],
       [0.-1.j]])

Para hacer un cambio de fase. Se puede ayudar con la puerta $RZ(\pi) RX(\pi) |0\rangle $

In [29]:
np.round(rz.dot(rx.dot(ket_0)), 5)

array([[0.-0.j],
       [1.-0.j]])

In [ ]:
from qiskit import QuantumCircuit
from qiskit.visualization import array_to_latex
from qiskit.quantum_info import Operator

qc = QuantumCircuit(1,1)
qc.rx(np.pi, 0)
operator = Operator(qc)
display(array_to_latex(operator, prefix="\\text{Circuito como matriz unitaria} = "))

<IPython.core.display.Latex object>

# Puertas U

In [ ]:
def get_matrix_u(theta, psi,  lambda ):
  return port

array([[0.+0.j],
       [0.-1.j]])

# Puertas para dos qubits

## Puerta CNOT